### PR TITLE
feature(cy): add anytrading env

### PR DESCRIPTION
**Pull request by Cloud-Pku**
## Description
Based on gym-anytrading, we create a stock environment for reinforcement learning-based trading algorithms.
1. For original DI-engine, modifications are made at [ding/worker/collector/interaction_serial_evaluator.py] as well as [ding/worker/collector/interaction_serial_evaluator.py]. These changes won't affect the usage of original DI-engine. The purpose of these changes is to print max profit in tensorboard.
2. Add the trading_test folder, which includes our stocks-v0 environnment. And stocks_dqn_config.py give a baseline based on DQN+stocks-v0.

## Check List

- [x] merge the latest version source branch/repo, and resolve all the conflicts
- [x] pass style check
- [x] pass all the tests

### CODE REVIEW

File: dizoo/trading_test/envs/trading_env.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+    def _calculate_reward(self, action):
+        raise NotImplementedError
+
+
+    def _update_profit(self, action):
+        raise NotImplementedError
+
+
+    def max_possible_profit(self):  # trade fees are ignored
```
Comment (Will-Nie) [inline dizoo/trading_test/envs/trading_env.py:226]: delete the comment

File: author.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+author:chenyun
```
Comment (PaParaZz1) [inline author.txt:1]: remove this file

File: ding/worker/collector/interaction_serial_evaluator.py [hunk 3]
```diff
@@ -187,6 +187,8 @@ def eval(
             - stop_flag (:obj:`bool`): Whether this training program can be ended.
```
Comment (PaParaZz1) [inline ding/worker/collector/interaction_serial_evaluator.py:187]: you can add a subclass of `InteractionSerialEvaluator` to implement the parts of code related to trading, and push into `dizoo/gym_anytrading/worker`

File: ding/worker/collector/interaction_serial_evaluator.py [hunk 4]
```diff
@@ -187,6 +187,8 @@ def eval(
...
             - return_info (:obj:`dict`): Current evaluation return information.
         '''
+        ### only used by anytrading
+        _anytrading_flag = False
```
Comment (PaParaZz1) [inline ding/worker/collector/interaction_serial_evaluator.py:191]: why starts with underline

File: dizoo/trading_test/config/calculate_max_profit.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+from easydict import EasyDict
```
Comment (PaParaZz1) [inline dizoo/trading_test/config/calculate_max_profit.py:1]: use pytest to implement unittest, you can get started from this [guideline](https://di-engine-docs.readthedocs.io/zh_CN/latest/22_test/index_zh.html)

File: dizoo/trading_test/config/stocks_dqn_config.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    exp_name='stocks_test_v9',
+    env=dict(
+        # Whether to use shared memory. Only effective if "env_manager_type" is 'subprocess'
+        # Env number respectively for collector and evaluator.
+        collector_env_num=8,
+        evaluator_env_num=8,
+        env_id='stocks-v0',
+        n_evaluator_episode=8,
+        stop_value=2,
+        # one year
```
Comment (PaParaZz1) [inline dizoo/trading_test/config/stocks_dqn_config.py:14]: worker days of one year?

File: dizoo/trading_test/envs/stocks_env.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        
+        
+        signal_features = np.column_stack((prices, diff, volumes))
+        # signal_features = np.column_stack((prices, opens, highs, lows, adjclose, volumes))
+        # signal_features = np.column_stack((prices, diff, opens, highs, lows, adjclose, volumes))
+        
+        return prices, signal_features
+
+
+    def _calculate_reward(self, action):
```
Comment (PaParaZz1) [inline dizoo/trading_test/envs/stocks_env.py:54]: add python typing lint

File: dizoo/trading_test/envs/stocks_env.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
+import numpy as np
+from .trading_env import TradingEnv, Actions, Positions
```
Comment (PaParaZz1) [inline dizoo/trading_test/envs/stocks_env.py:2]: polish import sequence

File: dizoo/trading_test/envs/trading_env.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    Buy = 3
+    Double_Buy = 4
+
+
+class Positions(Enum):
+    Short = -1.
+    Flat = 0.
+    Long = 1.
+
+def trans(position,  action):
```
Comment (PaParaZz1) [inline dizoo/trading_test/envs/trading_env.py:35]: rename to `transition` or `dynamics`?

File: dizoo/trading_test/envs/trading_env.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+
+def load_dataset(name, index_name):
+    base_dir = os.path.dirname(os.path.abspath(__file__))
+    path = os.path.join(base_dir, 'data', name + '.csv')
+    df = pd.read_csv(path, parse_dates=True, index_col=index_name)
+    return df
+
+class Actions(Enum):
+    Double_Sell = 0
```
Comment (PaParaZz1) [inline dizoo/trading_test/envs/trading_env.py:23]: we usually name enum element with all uppercase letter, such as `DOUBLE_SELL`

File: dizoo/trading_test/envs/trading_env.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+            self.plot_freq = 10
+        else:
+            self.plot_freq = self._cfg.plot_freq
+        if 'save_path' not in self._cfg :
+            self.save_path = './'
+        else:
+            self.save_path = self._cfg.save_path
+        #================================
+
+        STOCKS_GOOGL = load_dataset('STOCKS_GOOGL', 'Date')
```
Comment (PaParaZz1) [inline dizoo/trading_test/envs/trading_env.py:81]: you shouldn't add data in code repo, maybe you can add a download link here (ask for @norman26625 for more help)

File: dizoo/trading_test/envs/trading_env.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+        self.raw_prices = deepcopy(STOCKS_GOOGL).loc[:, 'Close'].to_numpy()
+        self.df = deepcopy(STOCKS_GOOGL).apply(lambda x: (x-x.mean())/ x.std(), axis=0) # normalize
```
Comment (PaParaZz1) [inline dizoo/trading_test/envs/trading_env.py:83]: add a eps in std

File: dizoo/trading_test/envs/stocks_env.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        step_reward = 0.
+        current_price = (self.raw_prices[self._current_tick])
+        last_trade_price = (self.raw_prices[self._last_trade_tick])
+        ratio = current_price/last_trade_price
+        cost = np.log((1 - self.trade_fee_ask_percent)*(1 - self.trade_fee_bid_percent))
+
+        if (action == Actions.Buy.value and self._position == Positions.Short):
+            step_reward = np.log(2-ratio) + cost
+        
+        if (action == Actions.Sell.value and self._position == Positions.Long):
```
Comment (PaParaZz1) [inline dizoo/trading_test/envs/stocks_env.py:64]: remove bracket here

File: dizoo/trading_test/config/calculate_max_profit.py (hunk 5 above)
Comment (PaParaZz1) [inline dizoo/trading_test/config/calculate_max_profit.py:1]: rename this directory from `dizoo/trading_test` to `dizoo/gym_anytrading`

File: dizoo/trading_test/envs/trading_env.py [hunk 14]
```diff
@@ -0,0 +1,243 @@
+from cmath import inf
```
Comment (PaParaZz1) [inline dizoo/trading_test/envs/trading_env.py:1]: add a readme to introduce our modification briefly

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/trading_test/config/stocks_dqn_config.py (hunk 6 above)
Comment (Cloud-Pku) [reply dizoo/trading_test/config/stocks_dqn_config.py]: 252 trading days a year

File: dizoo/gym_anytrading/envs/stocks_env.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
+from typing import Any
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/stocks_env.py:1]: correct code format in your modifications like `dizoo/gym_anytrading`, you can refer to [this](https://di-engine-docs.readthedocs.io/zh_CN/latest/21_code_style/index_zh.html)

File: dizoo/gym_anytrading/envs/stocks_env.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+        '''
+        Overview:
+            used by env.reset(), process the raw data.
+        Arguments:
+            - start_idx (int): the start tick; if None, then randomly select.
+        Returns:
+            - prices: the close.
+            - signal_features: feature map
+        '''
+        prices = self.df.loc[:, 'Close'].to_numpy()
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/stocks_env.py:29]: maybe we can implement this with a for loop:
```python
all_feature_name = ['a', 'b', 'c']
selected_feature_name = ['a', 'c']
all_feature = {k: self.df.loc[:, n].to_numpy() for n in all_feature_name]
selected_feature = np.column_stack([all_feature[n] for n in selected_feature_name])
```

File: dizoo/gym_anytrading/envs/trading_env.py [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+
+        plt.plot(short_ticks, eps_price[short_ticks], 'ro')
+        plt.plot(long_ticks, eps_price[long_ticks], 'go')
+        plt.plot(flat_ticks, eps_price[flat_ticks], 'bo')
+        plt.savefig(self.save_path + str(self._env_id) + '-price.png')
+
+    def close(self):
+        plt.close()
+
+    def _process_data(self):
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/trading_env.py:209]: you can use `ABC` and `abstractmethod` to restrict that subclasses must implement these three methods

File: dizoo/gym_anytrading/envs/trading_env.py [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+        if self._done:
+            if self._env_id[-1] == 'e' and self.cnt % self.plot_freq == 0:
+                self.render()
+            info['max_possible_profit'] = np.log(self.max_possible_profit())
+            info['final_eval_reward'] = self._total_reward
+
+        return BaseEnvTimestep(observation, step_reward, self._done, info)
+
+    def _get_observation(self):
+        obs = to_ndarray(self.signal_features[(self._current_tick - self.window_size + 1):self._current_tick + 1]
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/trading_env.py:174]: maybe you can split them into two lines

File: dizoo/gym_anytrading/envs/trading_env.py [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+        else:
+            self.plot_freq = self._cfg.plot_freq
+        if 'save_path' not in self._cfg:
+            self.save_path = './'
+        else:
+            self.save_path = self._cfg.save_path
+        #================================
+
+        STOCKS_GOOGL = load_dataset('STOCKS_GOOGL', 'Date')
+        self.raw_prices = deepcopy(STOCKS_GOOGL).loc[:, 'Close'].to_numpy()
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/trading_env.py:96]: maybe we can omit this deepcopy

File: dizoo/gym_anytrading/envs/trading_env.py [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+from gym.utils import seeding
+import numpy as np
+from enum import Enum
+import matplotlib.pyplot as plt
+from ding.envs import BaseEnv, BaseEnvTimestep
+from ding.utils import ENV_REGISTRY
+import os
+import pandas as pd
+from copy import deepcopy
+from ding.torch_utils import to_ndarray
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/trading_env.py:14]: polish import sequence as follows:
```python
from xxx import xxx
import xxx

from ding.xxx import xxx
from .xxx import xxx
```

File: dizoo/gym_anytrading/envs/stocks_env.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+        if start_idx is None:
+            self.start_idx = np.random.randint(self.window_size, len(self.df) - self._cfg.eps_length - 1)
+        else:
+            self.start_idx = start_idx
+
+        self._start_tick = self.start_idx
+        self._end_tick = self._start_tick + self._cfg.eps_length - 1
+
+        return prices, selected_feature
+
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/stocks_env.py:53]: add override comment:
```python
# override
def _calculate_reward(self, action: int) -> np.float32:
```

File: ding/entry/serial_entry_for_anytrading.py [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+        input_cfg: Union[str, Tuple[dict, dict]],
+        seed: int = 0,
+        env_setting: Optional[List[Any]] = None,
+        model: Optional[torch.nn.Module] = None,
+        max_train_iter: Optional[int] = int(1e10),
+        max_env_step: Optional[int] = int(1e10),
+) -> 'Policy':  # noqa
+    """
+    Overview:
+        Serial pipeline entry for off-policy RL.
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_for_anytrading.py:29]: modify comments

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/gym_anytrading/envs/stocks_env.py [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+        all_feature['Diff'] = diff
+        # =================================
+
+        # you can select features you want
+        selected_feature_name = ['Close', 'Diff', 'Volume']
+        selected_feature = np.column_stack([all_feature[k] for k in selected_feature_name])
+
+        # validate index
+        if start_idx is None:
+            self.start_idx = np.random.randint(self.window_size, len(self.df) - self._cfg.eps_length - 1)
```
Comment (lixl-st) [inline dizoo/gym_anytrading/envs/stocks_env.py:53]: the output of `randint(low, high)` falls into [low, high), maybe the right side is too restricted.

File: dizoo/gym_anytrading/envs/stocks_env.py [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+            step_reward = np.log(ratio) + cost
+
+        if action == Actions.DOUBLE_BUY.value and self._position == Positions.SHORT:
+            step_reward = np.log(2 - ratio) + cost
+
+        step_reward = to_ndarray([step_reward]).astype(np.float32)
+
+        return step_reward
+
+    def max_possible_profit(self) -> float:
```
Comment (lixl-st) [inline dizoo/gym_anytrading/envs/stocks_env.py:85]: [No need to change, just a quick discussion]

This function is fine with no transaction fees, as it takes full advantage of each stock price up and down
But when considering transaction fees, I will show you an example:
Given the transaction fee is `1$` on both long and short: 
T0: the price goes up `0.8$`;
T1: the price goes down `0.2$`;
T2: the price goes up `0.7$`.
According to the function, you should not buy or sell in this process.
However, it is possible to earn `0.8 - 0.2 + 0.7 - 1 = 0.3$`.

I just found the drawback but no perfect solution at hand, we can discuss this later.

File: dizoo/gym_anytrading/envs/trading_env.py [hunk 25]
```diff
@@ -0,0 +1,231 @@
...
+        self._done = None
+        self._current_tick = None
+        self._last_trade_tick = None
+        self._position = None
+        self._position_history = None
+        self._total_reward = None
+        #======================================
+
+        self._env_id = cfg.env_id
+        self._action_space = spaces.Discrete(len(Actions))
```
[the remark this answers is not in the document]
Comment (Cloud-Pku) [reply dizoo/gym_anytrading/envs/trading_env.py:113]: We get the action index(0~4) from DQN head, then we won't consider the data alignment anymore.

File: dizoo/gym_anytrading/envs/stocks_env.py (hunk 24 above)
Comment (Cloud-Pku) [reply dizoo/gym_anytrading/envs/stocks_env.py:91]: Actually, the calculation is based on ratio.
The calculation formula of profit is $\prod_{buying\ long}(r_{curr}/r_{pre}\ *\ cost) + \prod_{short\ selling}((2-r_{curr}/r_{pre})\ *\ cost)$. The profit is initialized to _1_ , which means 100%.
Take buying long as an example, the code shows that
> tmp_profit = profit * (current_price / last_trade_price) * (1 - self.trade_fee_ask_percent) * (1 - self.trade_fee_bid_percent)

T0: close is 212.557556;
T1: close is 215.340347;

transaction fee ratio is (1 - 0.01) * (1 - 0.005) = 0.98505
then 

T1_profit = T0_profit * (215.340347 / 212.557556) * 0.98505 
=  T0_profit * 1.013092 * 0.98505 
= T0_profit * 0.997946

In the case, although the closing price is rising, we should not take "buying long". 
And the code as following avoids this situation.
>  profit = max(profit, tmp_profit)

Comment (Cloud-Pku) [reply dizoo/gym_anytrading/envs/stocks_env.py:91]: Oh! I know what you mean. I'll reconsider it.

Comment (Cloud-Pku) [reply dizoo/gym_anytrading/envs/stocks_env.py:93]: This is a lower bound of _max_possible_profit_. Maybe we need to use dynamic programming to accurately calculate it.

**Review by PaParaZz1 — APPROVED** [review]

File: dizoo/gym_anytrading/envs/README.md [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+- otherwise:
+  - 0
+  
+so that maximize $\mathbb{E}_{\tau} \sum r$ 
+is equivalent to maximize $\mathbb{E}_{\tau}[\prod_{buying\ long}(r_{curr}/r_{pre}\ *\ cost) + \prod_{short\ selling}((2-r_{curr}/r_{pre})\ *\ cost)]$
+
+The experimental results show that such a defination is better than the original gym-anytrading accumulated reward function :$\sum(r_{curr} - r_{pre})$.
+### Render Function
+
+- As you see, you can use `render` method to plot the position and profit at one eposide.
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/README.md:84]: typo `eposide`

File: dizoo/gym_anytrading/envs/README.md [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+  
+
+![position](./position.png)![profit](./profit.png)
```
Comment (PaParaZz1) [inline dizoo/gym_anytrading/envs/README.md:87]: add x-axis and y-axis for these two images, and there may be a naive overview sentence to describe both figures.

**Review by PaParaZz1 — APPROVED** [review]
